src/common/helpers/snap-builds.js

```diff
@@ -25,43 +25,116 @@
+  'PUBLISH_NOW':     createState('Built, publishing now', BuildStatusColours.BLUE, 'tick', 3),
+  'PUBLISH_FAILED':  createState('Built, failed to publish', BuildStatusColours.RED, 'tick', 4),
+  'PUBLISHED':       createState('Built and published', BuildStatusColours.GREEN, 'tick_filled', 5)
+};
+
+function createState(message, colour, icon, priority) {
+  return {
+    message,
+    colour,
+    icon,
+    priority
+  };
+}
+
 // Based on BuildStatusConstants from LP API
 // https://git.launchpad.net/launchpad/tree/lib/lp/buildmaster/enums.py#n22
 //
-// mapping between build status from LP and red, yellow, green, grey status colours
-const BuildStatusMapping = {
-  'Needs building': BuildStatusColours.YELLOW,
-  'Successfully built': BuildStatusColours.GREEN,
-  'Failed to build': BuildStatusColours.RED,
-  'Dependency wait': BuildStatusColours.RED,
-  'Chroot problem': BuildStatusColours.RED,
-  'Build for superseded Source': BuildStatusColours.RED,
-  'Currently building': BuildStatusColours.YELLOW,
-  'Failed to upload': BuildStatusColours.RED,
-  'Uploading build': BuildStatusColours.YELLOW,
-  'Cancelling build': BuildStatusColours.RED,
-  'Cancelled build': BuildStatusColours.RED
+const LaunchpadBuildStates = {
+  'NEEDS': 'Needs building',
+  'SUCCESS': 'Successfully built',
+  'FAILED': 'Failed to build',
+  'WAIT': 'Dependency wait',
+  'CHROOT': 'Chroot problem',
+  'SUPERCEDED': 'Build for superseded Source',
+  'BUILDING': 'Currently building',
+  'UPLOAD_FAIL': 'Failed to upload',
+  'UPLOADING': 'Uploading build',
+  'CANCELLING': 'Cancelling build',
+  'CANCELLED': 'Cancelled build'
+};
+
+const LaunchpadStoreUploadStates = {
+  'UNSCHEDULED': 'Unscheduled',
+  'PENDING': 'Pending',
+  'FAILED_UPLOAD': 'Failed to upload',
+  'FAILED_RELEASE': 'Failed to release to channels',
+  'UPLOADED': 'Uploaded'
 };
 
+function mapBuildAndPublishedStates(buildState, publishState) {
+
+  if (buildState === LaunchpadBuildStates.SUCCESS) {
+    switch (publishState) {
+      case LaunchpadStoreUploadStates.UNSCHEDULED:
+        return BuildAndPubState.WONT_PUBLISH;
+
+      case LaunchpadStoreUploadStates.PENDING:
+        return BuildAndPubState.PUBLISH_SOON;
+
+      case LaunchpadStoreUploadStates.FAILED_UPLOAD:
+      case LaunchpadStoreUploadStates.FAILED_RELEASE:
+        return BuildAndPubState.PUBLISH_FAILED;
+
+      case LaunchpadStoreUploadStates.UPLOADED:
+        return BuildAndPubState.BUILT;
+    }
+  }
+
+  if (buildState === LaunchpadBuildStates.NEEDS) {
+    return BuildAndPubState.NEVER_BUILT;
+  }
+
+  if (buildState === LaunchpadBuildStates.FAILED) {
+    return BuildAndPubState.FAILED;
+  }
+
+  if (buildState === LaunchpadBuildStates.WAIT) {
+    return BuildAndPubState.SOON;
+  }
+
+  if (buildState === LaunchpadBuildStates.BUILDING) {
+    return BuildAndPubState.IN_PROGRESS;
+  }
+
+  if (buildState === LaunchpadBuildStates.UPLOADING) {
+    return BuildAndPubState.PUBLISH_NOW;
+  }
```

Comment:
There seem to be several other states not covered here yet.  As @bartaz said, it'd probably be clearer if it were `switch (buildState)` and then it could just cover all the possible states in their order of declaration, so that it's easy to see that none are left out.